float32 dist_f
float32 dist_b